lib/spack/spack/test/cmd/info.py

```diff
@@ -85,8 +50,29 @@ def test_info_fields(pkg_query, extra_args, parser, print_buffer):
         "Licenses:",
     )
 
-    args = parser.parse_args(["--all", pkg_query] + extra_args)
-    spack.cmd.info.info(parser, args)
+    output = info("--all", pkg_query, *extra_args)
+    assert all(field in output for field in expected_fields)
 
-    for text in expected_fields:
-        assert any(x for x in print_buffer if text in x)
+
+@pytest.mark.parametrize(
+    "by_name,args,in_output,not_in_output",
+    [
+        (True, ["licenses-1 +foo"], ["MIT"], ["Apache-2.0"]),
+        (True, ["licenses-1 ~foo"], ["Apache-2.0"], ["MIT"]),
+        (True, ["bowtie"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
+        (True, ["bowtie@1.2:"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
+        (True, ["bowtie@1.3:"], ["1.4.0", "1.3.0"], ["1.2.2", "1.2.0"]),
+        (True, ["bowtie@1.2"], ["1.2.2", "1.2.0"], ["1.3.0"]),  # 1.4.0 still shown as preferred
+        (False, ["licenses-1 +foo"], ["MIT"], ["Apache-2.0"]),
+        (False, ["licenses-1 ~foo"], ["Apache-2.0"], ["MIT"]),
+        (False, ["bowtie"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
+        (False, ["bowtie@1.2:"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
+        (False, ["bowtie@1.3:"], ["1.4.0", "1.3.0"], ["1.2.2", "1.2.0"]),
+        (False, ["bowtie@1.2"], ["1.2.2", "1.2.0"], ["1.3.0"]),  # 1.4.0 still shown as preferred
+    ],
+)
```

Comment:
If, as it seems, we expect no differences between tests using `--by-name` vs. not, we can shorten the definition like:
```suggestion
    "args,in_output,not_in_output",
    [
        (["licenses-1 +foo"], ["MIT"], ["Apache-2.0"]),
        (["licenses-1 ~foo"], ["Apache-2.0"], ["MIT"]),
        (["bowtie"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
        (["bowtie@1.2:"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
        (["bowtie@1.3:"], ["1.4.0", "1.3.0"], ["1.2.2", "1.2.0"]),
        (["bowtie@1.2"], ["1.2.2", "1.2.0"], ["1.3.0"]),  # 1.4.0 still shown as preferred
    ],
)
@pytest.mark.parametrize("by_name", [True, False])
```

Reply:
It would be good to also test single valued variants, and conditional variants. If there are cases where we would like to have a result, but it's difficult to infer it, I think it's fine to document the case as xfail.

Reply:
done in a testing commit I'm going to push